File: utils/expression-atlas/src/sdrf.rs
use anyhow::Result;
use polars::chunked_array::builder::Utf8ChunkedBuilder;
use polars::frame::DataFrame;
use polars::prelude::IntoSeries;
use std::fs;
use std::io::Read;
use std::path::PathBuf;

use log::{info, warn};

pub fn parse_condensed_sdrf(path: &PathBuf) -> Result<DataFrame, polars::prelude::PolarsError> {
    /*
    A condensed sdrf file has 7 columns, but the last is often not delimited correctly meaning it
    is tricky to read with the polars default csv reader.

    Therefore, we will be manually parsing the file into 6 series objects (one column seems to
    always be null) and constructing a dataframe from them

    We use a chunked array builder for Utf8 strings.
    */

    info!("Loading sdrf data from {:?}", path);

    let mut file = fs::File::open(path).unwrap();
    let mut s = String::new();
    file.read_to_string(&mut s)?;

    let part_parsed: Vec<Vec<&str>> = s.lines().map(|line| line.split('\t').collect()).collect();
    let bytes_per_string: usize = 128;
    let mut exp_name = Utf8ChunkedBuilder::new("exp_name", part_parsed.len(), bytes_per_string);
    let mut assay_name = Utf8ChunkedBuilder::new("assay_name", part_parsed.len(), bytes_per_string);
    let mut feat_class = Utf8ChunkedBuilder::new("feat_class", part_parsed.len(), bytes_per_string);
    let mut feat_type = Utf8ChunkedBuilder::new("feat_type", part_parsed.len(), bytes_per_string);
    let mut feat_value = Utf8ChunkedBuilder::new("feat_value", part_parsed.len(), bytes_per_string);
    let mut ontology = Utf8ChunkedBuilder::new("ontology", part_parsed.len(), bytes_per_string);

    // There is one experiment file that does not have the empty column in line[1]
    if part_parsed.iter().map(|x| x.len()).max().unwrap() == 7 {
        for line in part_parsed.iter() {
            exp_name.append_value(line[0]);
            assay_name.append_value(line[2]); // remember line[1] will be empty
            feat_class.append_value(line[3]);
            feat_type.append_value(line[4]);
            feat_value.append_value(line[5]);
            if line.len() == 7 {
                ontology.append_value(line[6]);
            } else {
                ontology.append_null();
            }
        }
    } else {
        warn!("Unusual sdrf parsing with {} columns, not 7 for experiment {}",part_parsed.iter().map(|x| x.len()).max().unwrap(), part_parsed[0][0]);
        for line in part_parsed.iter() {
            exp_name.append_value(line[0]);
            assay_name.append_value(line[1]);
            feat_class.append_value(line[2]);
            feat_type.append_value(line[3]);
            feat_value.append_value(line[4]);
            if line.len() == 6 {
                ontology.append_value(line[5]);
            } else {
                ontology.append_null();
            }
        }
    }

    // Now have all the lines parsed with the same lengths. Try to construct a dataframe...
    DataFrame::new(vec![
        exp_name.finish().into_series(),
        assay_name.finish().into_series(),
        feat_class.finish().into_series(),
        feat_type.finish().into_series(),
        feat_value.finish().into_series(),
        ontology.finish().into_series(),
    ])
}

File: rnacentral_pipeline/databases/expressionatlas/sdrf.py
import logging
import os
from pathlib import Path
from typing import List, Union

import polars as pl


def parse_condensed_sdrf(path: Union[str, Path]) -> pl.DataFrame:
    """
    Parse a condensed SDRF file and return a polars DataFrame.

    A condensed SDRF file has 7 columns, but the last is often not delimited correctly,
    making it tricky to read with the polars default CSV reader.
    Therefore, manually parse the file into 6 series objects (one column seems to
    always be null) and construct a dataframe from them.

    Args:
        path: Path to the SDRF file

    Returns:
        A polars DataFrame containing the parsed SDRF data
    """
    logging.info(f"Loading sdrf data from {path}")

    # Read the file
    with open(path, "r") as file:
        content = file.read()

    # Parse lines and split by tabs
    part_parsed: List[List[str]] = [
        line.strip().split("\t") for line in content.splitlines()
    ]

    if len(part_parsed) == 0:
        raise ValueError(f"SDRF file {path} contained no data")
    # Initialize empty lists for each column
    exp_name_data = []
    assay_name_data = []
    feat_class_data = []
    feat_type_data = []
    feat_value_data = []
    ontology_data = []

    # Get the maximum number of columns
    max_columns = max(len(line) for line in part_parsed)

    # Parse according to the number of columns
    if max_columns == 7:
        for line in part_parsed:
            exp_name_data.append(line[0])
            assay_name_data.append(line[2])  # remember line[1] will be empty
            feat_class_data.append(line[3])
            feat_type_data.append(line[4])
            feat_value_data.append(line[5])
            if len(line) == 7:
                ontology_data.append(line[6])
            else:
                ontology_data.append(None)
    else:
        logging.warning(
            f"Unusual sdrf parsing with {max_columns} columns, not 7 for experiment {part_parsed[0][0]}"
        )
        for line in part_parsed:
            exp_name_data.append(line[0])
            assay_name_data.append(line[1])
            feat_class_data.append(line[2])
            feat_type_data.append(line[3])
            feat_value_data.append(line[4])

            if len(line) == 6:
                ontology_data.append(line[5])
            else:
                ontology_data.append(None)

    # Create the DataFrame with all series
    df = pl.DataFrame(
        {
            "exp_name": exp_name_data,
            "assay_name": assay_name_data,
            "feat_class": feat_class_data,
            "feat_type": feat_type_data,
            "feat_value": feat_value_data,
            "ontology": ontology_data,
        }
    )

    return df
